peps/pep-0807.rst

```diff
@@ -0,0 +1,398 @@
+Author: William Woodruff <william@yossarian.net>
+Sponsor: Donald Stufft <donald@stufft.io>
+PEP-Delegate: Donald Stufft <donald@stufft.io>
+Status: Draft
+Type: Standards Track
+Topic: Packaging
+Created: 15-Sep-2025
+Post-History: `08-Aug-2025 <https://discuss.python.org/t/pre-pep-trusted-publishing-token-exchange/103067>`__
+
+Abstract
+========
+
+This PEP proposes a standard mechanism through which arbitrary
+Python package indices can support "Trusted Publishing," a misuse-resistant
+credential exchange scheme already implemented by the Python Package Index
+(PyPI).
+
+The mechanism proposed in this PEP is designed to encapsulate PyPI's
+`existing implementation <https://docs.pypi.org/trusted-publishers/>`_
+of Trusted Publishing, while allowing other indices to implement the same
+scheme in a manner that is discoverable by and interoperable with existing
+Python package uploading clients.
+
+Rationale and Motivation
+========================
+
+"Trusted Publishing" is PyPI's term of art for using the
+`OpenID Connect (OIDC) standard <https://openid.net/connect/>`_
+to exchange a short-lived *identity credential* from a trusted
+third-party service (like a CI/CD or cloud provider) for a short-lived,
+minimally-scoped *upload credential* that can be used to publish
+to the index.
+
+Trusted Publishing was originally designed and enabled on PyPI in 2023 as
+a non-standard (PyPI-specific) feature, much like the existing
+`upload API <https://docs.pypi.org/trusted-publishers/>`_. It has seen
+widespread adoption in that capacity: over 640,000 files have been published
+to PyPI using a Trusted Publisher (as of September 2025), and PyPI's
+design has inspired similar designs in the
+`Rust (crates.io) <https://crates.io/docs/trusted-publishing>`_,
+`Ruby (RubyGems) <https://guides.rubygems.org/trusted-publishing/>`_, and
+`JavaScript (npm) <https://docs.npmjs.com/trusted-publishers>`_ ecosystems.
+
+The absence of a standard for Trusted Publishing presents a long-term
+impediment for adoption: third-party indices (i.e. those other than
+PyPI and TestPyPI) cannot easily implement Trusted Publishing without
+referencing PyPI's unstandardized design. This in turn poses a long-term
+maturity risk similar to that of the unstandardized upload API: package upload
+clients (like `twine <https://twine.readthedocs.io/en/stable/>`_ and
+`uv <https://docs.astral.sh/uv/>`_) must either accept behavioral differences
+between indices (leading to an accretion of hacks) or continue to reject
+non-PyPI implementations of Trusted Publishing.
+
+Specification
+=============
+
+This PEP's specification contains two parts:
+
+* A *discovery* mechanism that package upload clients can use to determine
+  whether an arbitrary Python package index host supports Trusted Publishing.
+* A *token exchange* mechanism that package upload clients can use to
+  exchange an identity credential for an upload credential.
+
+.. _constraints:
```

Comment:
This reference target doesn't seem to be used? If you are using it, please prefix it with the PEP number as all PEPs share a global namespace for link targets.
```suggestion
```